### ISSUE TITLE
Requests for images and other non-html content via a website endpoint not handled

### ISSUE THREAD

**Issue by jagregory** [jamhall/s3rver#502]
The website middleware doesn't handle any requests that don't preference `text/html` in their `Accept` header, which means requests for objects in the bucket that aren't HTML fail.

For example, when a browser makes a request for an image from an `img` element it uses a different `Accept` header than if you open the URL directly (same applies for `video` etc...). MDN has a list of default Accept headers here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation/List_of_default_Accept_values

In short, the current implementation fails to correctly respond to requests for images/videos/other content when the browser initiates the request not the user.

There's a failing test in my fork (branch: website-images): https://github.com/jagregory/s3rver/tree/website-images

**Comment by jagregory** [comment]
Any thoughts on what would be the best way to tackle this @kherock? I'm happy to send a PR with some guidance.

**Comment by kherock** [comment]
I'm also suspicious of the condition there
https://github.com/jamhall/s3rver/blob/183977983274468a9e6f514ce33db7e8a7120e7e/lib/middleware/website.js#L39-L46
but I haven't had a chance to test anything. The goal was to detect when the SDK loads a path so that it wouldn't perform any website behavior that would break it. If I can find an hour to spend stepping through, I could probably track down where it's going wrong. I feel like this is probably a simple fix.

**Comment by kherock** [comment]
@jagregory I tried running the test case you had in your fork, and it was failing because it was printing a ListBucketResult instead of requesting the actual key. When I updated the request to use ``uri: `${buckets[0].name}/image.png` ``, things seemed to pass. I also tried hosting it and testing against an actual browser. Let me know if I'm missing something.

**Comment by jagregory** [comment]
My testcase is obviously wrong. That's annoying. I'm able to replicate the issue locally by:

1. Having a WebsiteConfiguration that redirects requests (e.g. `/thumbnail/image.png` is redirected to `http://example.com/resize?key=image.png`
2. Uploading an image (e.g. to image.png)
3. Requesting the URL that would redirect (e.g. `/large/image.png)

If you do step 3 with an `Accept` with anything other than `text/html` the redirect doesn't occur and you're given the standard `NoSuchKey` response.

I'll try to put a better testcase together

**Comment by jagregory** [comment]
Actually you can simplify that quite a bit. For testing you don't need the redirect to actually go anywhere (as long as you get the status code. So it becomes:

1. Having a WebsiteConfiguration that redirects requests
2. Requesting the URL that would redirect with an Accept other than `text/html`

**Comment by jagregory** [comment]
Ok, testcase updated. Hope that's more helpful. Sorry about that.

**Comment by kherock** [comment]
No worries, thanks for clarifying the issue. I think I see the problem now. I need to update the Accept header check to have better SDK detection. Basically, I'd like it to skip the website middleware if S3rver detects that the client accepts application/xml content _but only after exhausing the acceptable types a browser lists before `*/*`_.

The existing behavior with `ctx.accept()` matches text/html since it's the first type a browser normally sends when loading a web page. For images and videos, the Accept header is typically just `*/*` for some browsers, so we might not be able to change the condition to something like
```javascript
ctx.accepts('application/xml', 'text/html', 'image', 'audio', 'video') === 'application/xml'
```
and be done with it. Though this is pretty close. I'll need to play around with `ctx.accepts()` to see if I can find something that works.

**Comment by jagregory** [comment]
What about instead of checking `Accept`, you look for a header only the AWS SDK and/or CLI sends? Perhaps an `x-amz` header of some kind?

**Comment by kherock** [comment]
The only solid evidence that the SDK is making the request (for the JavaScript SDK at least) is that the user agent is set to `aws-sdk-nodejs/2.507.0 win32/v8.15.0 promise` for my environment. The `Accept` header is omitted, but in the web version of the SDK, all requests are sent with `Accept: */*`.

**Comment by jagregory** [comment]
Any other thoughts on how we could support this? I'm hoping to use this to showcase the typical image thumbnail generator S3/Lambda use-case with s3rver, but it's a bit unsatisfying if you have to do it all through curl.

**Comment by jagregory** [comment]
I've done some research. I proxied requests through to S3 and used three different clients: the AWS CLI, AWS NodeJS SDK, and the Python Boto3 library. You can see the results below. All three consistently send the `X-Amz-Date` and `X-Amz-Content-Sha256`, and they also all set the `Authorization` header to a common pattern. These are all part of the [AWS request signing process](https://docs.aws.amazon.com/general/latest/gr/sigv4_elements.html) and I think we'd be fairly safe treating the presence of any `x-amz-*` headers as proof of an cli/sdk.

I'm going to give this a shot in a PR.

---

**boto3**
<img width="669" alt="aws-boto3" src="https://user-images.githubusercontent.com/10828/63992488-63d02e80-cb2f-11e9-8246-ab1c49c80279.png">

**aws-cli**
<img width="665" alt="aws-cli" src="https://user-images.githubusercontent.com/10828/63992489-63d02e80-cb2f-11e9-9c2f-0a7a0f6f5a38.png">

**node sdk**
<img width="640" alt="aws-nodesdk" src="https://user-images.githubusercontent.com/10828/63992490-63d02e80-cb2f-11e9-9a53-2f359a532155.png">
